### PR TITLE
fix(sql): race in parallel group by with many groups

### PR DESCRIPTION
**Pull request by puzpuzpuz**
When a parallel group by sharding takes place, it may happen that a work-stealing thread steals another thread's task and starts executing it as if it's the task's owner.

The fix makes the per-worker synchronization mandatory in such cases by checking the task's atom and comparing it with the thread's one.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 25 / 25 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByNotKeyedRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByMergeShardJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilterAtom.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByAtom.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByNotKeyedAtom.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|1|1|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByIntHashSet.java [hunk 1]
```diff
@@ -181,7 +182,9 @@ private long probe(int key, long index) {
             if (key == k) {
                 return -index - 1;
             }
-        } while (true);
+        } while (index != index0);
+
+        throw CairoException.critical(0).put("corrupt hash table");
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/groupby/GroupByIntHashSet.java:187]:
```suggestion
        throw CairoException.critical(0).put("corrupt integer hash set state");
```

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong128HashSet.java [hunk 2]
```diff
@@ -187,7 +188,9 @@ private long probe(long lo, long hi, long index) {
             if (loKey == lo && hiKey == hi) {
                 return -index - 1;
             }
-        } while (true);
+        } while (index != index0);
+
+        throw CairoException.critical(0).put("corrupt hash table");
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong128HashSet.java:193]:
```suggestion
        throw CairoException.critical(0).put("corrupt long128 hash set state");
```

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong256HashSet.java [hunk 3]
```diff
@@ -195,7 +196,9 @@ private long probe(long k0, long k1, long k2, long k3, long index) {
             if (k0Key == k0 && k1Key == k1 && k2Key == k2 && k3Key == k3) {
                 return -index - 1;
             }
-        } while (true);
+        } while (index != index0);
+
+        throw CairoException.critical(0).put("corrupt hash table");
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong256HashSet.java:201]:
```suggestion
        throw CairoException.critical(0).put("corrupt long256 hash set state");
```

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLongHashSet.java [hunk 4]
```diff
@@ -181,7 +182,9 @@ private long probe(long key, long index) {
             if (key == k) {
                 return -index - 1;
             }
-        } while (true);
+        } while (index != index0);
+
+        throw CairoException.critical(0).put("corrupt hash table");
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLongHashSet.java:187]:
```suggestion
        throw CairoException.critical(0).put("corrupt long64 hash set state");
```

File: core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java [hunk 5]
```diff
@@ -116,6 +104,18 @@ public void initCursor() {
...
 
+    public int maybeAcquireFilter(int workerId, boolean owner, SqlExecutionCircuitBreaker circuitBreaker) {
+        if (perWorkerLocks == null) {
+            return -1;
+        }
+        if (workerId == -1 && owner) {
+            // Owner thread is free to use its own filter anytime.
+            return -1;
+        }
+        // All other threads, e.g. worker or work stealing threads, must always acquire a lock.
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java:115]:
```suggestion
        // All other threads, e.g. worker or work stealing threads, must always acquire a lock to use shared resources.
```

File: core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java [hunk 6]
```diff
@@ -116,6 +104,18 @@ public void initCursor() {
         }
     }
 
+    public int maybeAcquireFilter(int workerId, boolean owner, SqlExecutionCircuitBreaker circuitBreaker) {
+        if (perWorkerLocks == null) {
+            return -1;
+        }
+        if (workerId == -1 && owner) {
+            // Owner thread is free to use its own filter anytime.
```
Comment (ideoma) [inline core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java:112]:
```suggestion
            // Owner thread is free to use its own private filter anytime.
```

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByIntHashSet.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/groupby/GroupByIntHashSet.java]: Changed in 20ba60c6b4d95ff2aec27a0f99f028032b891afe

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong128HashSet.java (hunk 2 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong128HashSet.java]: Changed in 20ba60c6b4d95ff2aec27a0f99f028032b891afe

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong256HashSet.java (hunk 3 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLong256HashSet.java]: Changed in 20ba60c6b4d95ff2aec27a0f99f028032b891afe

File: core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLongHashSet.java (hunk 4 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/groupby/GroupByLongHashSet.java]: Changed in 20ba60c6b4d95ff2aec27a0f99f028032b891afe

File: core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java (hunk 5 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java]: Changed in 20ba60c6b4d95ff2aec27a0f99f028032b891afe

File: core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java (hunk 6 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/griffin/engine/table/AsyncFilterAtom.java]: Changed in 20ba60c6b4d95ff2aec27a0f99f028032b891afe

**Review by ideoma — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 41 (80.49%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByLongHashSet.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByLong128HashSet.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByIntHashSet.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByLong256HashSet.java|2|4|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByNotKeyedRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByMergeShardJob.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilterAtom.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByAtom.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncJitFilteredRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursor.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByNotKeyedAtom.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncFilteredRecordCursorFactory.java|1|1|100.00%|

**Pull request merged**